common/thorhelper/thorfile.cpp

```diff
@@ -100,3 +101,257 @@ IHThorDiskReadArg * createWorkUnitReadArg(const char * filename, IHThorWorkunitR
+            offset_t thissize = ifile->size();
+            if (thissize != (offset_t)-1)
+            {
+                StringBuffer remotePath;
+                rfn.getPath(remotePath);
+                unsigned crc = 0;
+                keyFile.getCrc(crc);
+                return createKeyIndex(remotePath.str(), crc, false);
+            }
+        }
+        catch (IException *E)
+        {
+            EXCLOG(E, "While opening index file");
+            if (exc)
+                E->Release();
+            else
+                exc.setown(E);
+            failcount++;
+        }
+    }
+    if (exc)
+        throw exc.getClear();
+    StringBuffer url;
+    RemoteFilename rfn;
+    keyFile.getFilename(rfn).getRemotePath(url);
+    throw MakeStringException(1001, "Could not open key file at %s%s", url.str(), (numCopies > 1) ? " or any alternate location." : ".");
+}
+
+
+//-----------------------------------------------------------------------------
+
+constexpr unsigned defaultTimeout = 10000;
+
+static bool checkIndexMetaInformation(IDistributedFilePart * part, bool force)
+{
+    if (part->queryAttributes().hasProp("@offsetBranches"))
+        return true;
+
+    if (!force)
+        return false;
+
+    Owned<IKeyIndex> index = openKeyFile(*part);
+    offset_t branchOffset = index->queryFirstBranchOffset();
+
+    part->lockProperties(defaultTimeout);
+    part->queryAttributes().setPropInt64("@offsetBranches", branchOffset);
+    part->unlockProperties();
+    return true;
+}
+
+bool checkIndexMetaInformation(IDistributedFile * file, bool force)
+{
+    IDistributedSuperFile * super = file->querySuperFile();
+    if (super)
+    {
+        Owned<IDistributedFileIterator> subfiles = super->getSubFileIterator(true);
+        bool first = false;
+        ForEach(*subfiles)
+        {
+            IDistributedFile & cur = subfiles->query();
+            if (!checkIndexMetaInformation(&cur, force))
+                return false;
+        }
+        return true;
+    }
+
+    try
+    {
+        IPropertyTree & attrs = file->queryAttributes();
+        if (!attrs.hasProp("@nodeSize") || !attrs.hasProp("@keySize"))
+        {
+            if (!force)
+                return false;
+
+            //Read values from the header of the first index part and save in the meta data
+            IDistributedFilePart & part = file->queryPart(0);
+            Owned<IKeyIndex> index = openKeyFile(part);
+            size_t keySize = index->keySize();
+            size_t nodeSize = index->getNodeSize();
+
+            //FUTURE: When file information is returned from an esp service, will this be updated?
+            file->lockProperties(defaultTimeout);
+            file->queryAttributes().setPropInt("@nodeSize", nodeSize);
+            file->queryAttributes().setPropInt("@keySize", keySize);
+            file->unlockProperties();
+        }
+
+        if (!attrs.hasProp("@numLeafNodes"))
```

Comment:
question: should somewhere be setting this yet (e.g. in the case of no blob nodes) ?

Reply:
It is set by hthor and roxie when the index is built - there is a jira to add it to thor.  I think it is better not to set it here - it can deduce most of it from offsetBranches and it is useful to know it is an approximation.